### ISSUE TITLE
Webpack 4 outputs deprecation warnings

### ISSUE THREAD

**Issue by ai** [webpack-contrib/webpack-bundle-analyzer#154]
#### Issue description

Using `webpack-bundle-analyzer` with Webpack 4:

```
DepreciationWarning: Tappable.plugin is deprecated. Use new API on `.hooks` instead
```

#### Technical info
* Webpack Bundle Analyzer version: 2.10.0
* Webpack version: Webpack 4.0.0-beta.2
* Node.js version: 9
* npm/yarn version: 1.3.2
* OS: Linux

#### Debug info

How do you use this module? As CLI utility or as plugin?

Plugin inside [Size Limit](https://github.com/ai/size-limit/)

If plugin, what options were provided? (e.g. `new BundleAnalyzerPlugin({ analyzerMode: 'disabled', generateStatsFile: true })`)

```
      openAnalyzer: true,
      analyzerMode: 'server',
      defaultSizes: 'gzip'
```

What other Webpack plugins were used?

`compression-webpack-plugin` and `uglifyjs-webpack-plugin`

Related issue: https://github.com/ai/size-limit/pull/42

**Comment by valscion** [comment]
Hi, and thanks for raising an issue! So to clarify, `webpack-bundle-analyzer` works but it outputs a deprecation warning with webpack 4?

The deprecation warning is a result of this line:

https://github.com/webpack-contrib/webpack-bundle-analyzer/blob/b2a617c176f2fd76a12fa8806f3383867b148c1a/src/BundleAnalyzerPlugin.js#L35

Do you have an idea of how a plugin can support both webpack 3 and 4 at the same time without outputting deprecation warnings?

**Comment by ai** [comment]
@valscion 

> So to clarify, webpack-bundle-analyzer works but it outputs a deprecation warning with webpack 4?

Yeap :)

> Do you have an idea of how a plugin can support both webpack 3 and 4 at the same time

In the same cases, I used this method:

```js
if (compiler.hooks) {
  // Webpack >= 4
} else {
  // Webpack < 4
}
```

But I am not 100% sure about this case.

**Comment by valscion** [comment]
Should be pretty easy to do that :). Would you be open to trying this approach out with webpack 4? https://github.com/webpack-contrib/webpack-bundle-analyzer/blob/master/CONTRIBUTING.md has intructions on how to setup the build pipeline for this project and link to your own project for local testing. Or maybe @jaydenseric would like to give this a try if you've got time?

**Comment by ai** [comment]
@valscion sure, if nobody closes it until Friday, I will try to fix (Friday is my open source day =^_^=).

**Comment by valscion** [comment]
:+1: thanks!

I noticed that CONTRIBUTING.md still has a mention of `dev` branch, which to my recollection has been deleted from this repository quite some time ago already. #155 hopefully clarifies the situation a bit :relaxed:

**Comment by evilebottnawi** [comment]
@valscion better do major release than use `if (compiler.hooks) {} else {}`, we will do this in contrib after webpack 4 was release, right now is it just warning and it is not critical

**Comment by ai** [comment]
@evilebottnawi if you will do major release all next Bundle Analyzer releases will be unavailable for Webpack 3 users

**Comment by valscion** [comment]
> @valscion better do major release than use if (compiler.hooks) {} else {}, we will do this in contrib after webpack 4 was release, right now is it just warning and it is not critical

If possible @evilebottnawi, I'd rather support both webpack 3 and 4. We only use the `done` hook in this plugin so compared to other plugins and loaders, supporting both versions should be trivial. Does this sound like a good approach to you?

**Comment by th0r** [comment]
@evilebottnawi is there a recommended way to support both Webpack 4 and 3 APIs? @sokra?

**Comment by michael-ciniawsky** [comment]
`webpack >= 4` supports the 'old' Plugin API, so I recommend the compat is handled in/by webpack itself. There should simply be a flag or the like to suppress the warnings instead of working around that  by plugin/loader authors :). It's still just a warning and the code is functional

**Comment by sokra** [comment]
you can do the following:

```js
if(compiler.hooks) {
  compiler.hooks.done.tap(...);
} else {
  compiler.plugin(...);
}
```

**Comment by ai** [comment]
@valscion I made a PR #156

**Comment by valscion** [comment]
Fixed in v2.10.1 :tada:, thanks @ai for the pull request!
